File: server/app/api/runs.py
from __future__ import annotations

import asyncio
import json

from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import StreamingResponse

from server.app.application.runs import DraftNotReadyError, RunConflictError, RunConfigurationError, RunService
from server.app.models.run_session import CreateRunRequest, RunLogChunk, RunLogPreview, RunSession


def build_runs_router(service: RunService) -> APIRouter:
    router = APIRouter(tags=["runs"])

    @router.post("/runs", response_model=RunSession, status_code=status.HTTP_201_CREATED)
    def create_run(request: CreateRunRequest) -> RunSession:
        try:
            run = service.create_run(request)
        except DraftNotReadyError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        except RunConfigurationError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        if run is None:
            raise HTTPException(status_code=404, detail="Course draft not found")
        return run

    @router.get("/runs/{run_id}", response_model=RunSession)
    def get_run(run_id: str) -> RunSession:
        run = service.get_run(run_id)
        if run is None:
            raise HTTPException(status_code=404, detail="Run not found")
        return run

    @router.get("/runs/{run_id}/log", response_model=RunLogPreview)
    def get_run_log(run_id: str) -> RunLogPreview:
        payload = service.get_run_log(run_id)
        if payload is None:
            raise HTTPException(status_code=404, detail="Run not found")
        return payload

    @router.get("/runs/{run_id}/log/events")
    async def stream_run_log_events(run_id: str, cursor: int = Query(0, ge=0)):
        run = service.get_run(run_id)
        if run is None:
            raise HTTPException(status_code=404, detail="Run not found")

        async def event_stream():
            current_cursor = cursor
            while True:
                chunk = service.get_run_log_chunk(run_id, cursor=current_cursor)
                if chunk is None:
                    break
                current_cursor = chunk.cursor
                if chunk.content or chunk.complete:
                    payload = json.dumps(chunk.model_dump(), ensure_ascii=False)
                    # Pad with 2KB of comment to bypass uvicorn/proxy ASGI buffering
                    padding = ": " + " " * 2048 + "\n"
                    yield f"{padding}event: run.log\ndata: {payload}\n\n"
                if chunk.complete:
                    break
                await asyncio.sleep(1)

        return StreamingResponse(
            event_stream(),
            media_type="text/event-stream",
            headers={"Cache-Control": "no-cache"},
        )

    @router.post("/runs/{run_id}/resume", response_model=RunSession)
    def resume_run(run_id: str) -> RunSession:
        try:
            run = service.resume_run(run_id)
        except DraftNotReadyError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        except RunConfigurationError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        except RunConflictError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        if run is None:
            raise HTTPException(status_code=404, detail="Run not found")
        return run

    @router.post("/runs/{run_id}/clean", response_model=RunSession)
    def clean_run(run_id: str) -> RunSession:
        try:
            run = service.clean_run(run_id)
        except RunConflictError as error:
            raise HTTPException(status_code=409, detail=str(error)) from error
        if run is None:
            raise HTTPException(status_code=404, detail="Run not found")
        return run

    @router.get("/runs/{run_id}/events")
    async def stream_run_events(run_id: str):
        run = service.get_run(run_id)
        if run is None:
            raise HTTPException(status_code=404, detail="Run not found")

        async def event_stream():
            while True:
                current = service.get_run(run_id)
                if current is None:
                    break
                payload = json.dumps(current.model_dump(), ensure_ascii=False)
                # Pad with 2KB of comment to bypass uvicorn/proxy ASGI buffering
                padding = ": " + " " * 2048 + "\n"
                yield f"{padding}event: run.update\ndata: {payload}\n\n"
                if current.status in {"completed", "failed", "cleaned"}:
                    break
                await asyncio.sleep(1)

        return StreamingResponse(
            event_stream(),
            media_type="text/event-stream",
            headers={"Cache-Control": "no-cache"},
        )

    return router

File: web/lib/api/runs.ts
export type StageStatus = {
  name: string;
  status: string;
};

export type ChapterProgress = {
  chapter_id: string;
  status: string;
  current_step: string | null;
  completed_step_count: number;
  total_step_count: number;
  export_ready: boolean;
};
export type CourseResultsContext = {
  course_id: string;
  latest_run: RunSession | null;
};


export type RunSession = {
  id: string;
  draft_id: string;
  course_id: string;
  created_at: string | null;
  status: string;
  run_kind: "chapter" | "global";
  backend: string;
  hosted: boolean;
  base_url: string | null;
  simple_model: string | null;
  complex_model: string | null;
  timeout_seconds: number | null;
  target_output: string | null;
  review_enabled: boolean;
  review_mode: string | null;
  stages: StageStatus[];
  chapter_progress: ChapterProgress[];
  snapshot_complete: boolean;
  last_error: string | null;
};

export type UnstartedRunWorkbenchState = {
  draft_id: string | null;
  course_id: string | null;
};

export type RunLogPreview = {
  run_id: string;
  available: boolean;
  cursor: number;
  content: string;
  truncated: boolean;
};

export type RunLogChunk = {
  run_id: string;
  cursor: number;
  content: string;
  complete: boolean;
};

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://127.0.0.1:8000";

export async function createRun(
  draftId: string,
  options?: {
    review_enabled?: boolean;
    run_kind?: "chapter" | "global";
  },
): Promise<RunSession> {
  const response = await fetch(`${API_BASE_URL}/runs`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      draft_id: draftId,
      review_enabled: options?.review_enabled,
      run_kind: options?.run_kind ?? "chapter",
    }),
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to create run: ${response.status} ${message}`);
  }

  return (await response.json()) as RunSession;
}

export async function getRun(runId: string): Promise<RunSession> {
  const response = await fetch(`${API_BASE_URL}/runs/${runId}`, {
    cache: "no-store",
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to load run: ${response.status} ${message}`);
  }

  return (await response.json()) as RunSession;
}

export async function resumeRun(runId: string): Promise<RunSession> {
  const response = await fetch(`${API_BASE_URL}/runs/${runId}/resume`, {
    method: "POST",
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to resume run: ${response.status} ${message}`);
  }

  return (await response.json()) as RunSession;
}

export async function cleanRun(runId: string): Promise<RunSession> {
  const response = await fetch(`${API_BASE_URL}/runs/${runId}/clean`, {
    method: "POST",
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to clean run: ${response.status} ${message}`);
  }

  return (await response.json()) as RunSession;
}

export async function getCourseResultsContext(courseId: string): Promise<CourseResultsContext> {
  const response = await fetch(`${API_BASE_URL}/courses/${courseId}/results-context`, {
    cache: "no-store",
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to load course results context: ${response.status} ${message}`);
  }

  return (await response.json()) as CourseResultsContext;
}

export async function getRunLog(runId: string): Promise<RunLogPreview> {
  const response = await fetch(`${API_BASE_URL}/runs/${runId}/log`, {
    cache: "no-store",
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to load run log: ${response.status} ${message}`);
  }

  return (await response.json()) as RunLogPreview;
}

export function subscribeRunLogEvents(
  runId: string,
  cursor: number,
  handlers: {
    onChunk: (chunk: RunLogChunk) => void;
    onError?: (message: string) => void;
  },
): () => void {
  const eventSource = new EventSource(`${API_BASE_URL}/runs/${runId}/log/events?cursor=${cursor}`);
  let closed = false;

  const handleChunk = (event: MessageEvent<string>) => {
    try {
      const payload = JSON.parse(event.data) as RunLogChunk;
      handlers.onChunk(payload);
      if (payload.complete) {
        closed = true;
        eventSource.close();
      }
    } catch (error) {
      handlers.onError?.(error instanceof Error ? error.message : "Failed to parse run log event");
    }
  };

  const handleError = () => {
    if (!closed) {
      handlers.onError?.("Run log stream disconnected");
    }
  };

  eventSource.addEventListener("run.log", handleChunk as EventListener);
  eventSource.onerror = handleError;

  return () => {
    closed = true;
    eventSource.removeEventListener("run.log", handleChunk as EventListener);
    eventSource.close();
  };
}

export function subscribeRunEvents(
  runId: string,
  handlers: {
    onUpdate: (run: RunSession) => void;
    onError?: (message: string) => void;
  },
): () => void {
  const eventSource = new EventSource(`${API_BASE_URL}/runs/${runId}/events`);
  let closed = false;

  const handleUpdate = (event: MessageEvent<string>) => {
    try {
      const payload = JSON.parse(event.data) as RunSession;
      handlers.onUpdate(payload);
      if (payload.status === "completed" || payload.status === "failed" || payload.status === "cleaned") {
        closed = true;
        eventSource.close();
      }
    } catch (error) {
      handlers.onError?.(error instanceof Error ? error.message : "Failed to parse run event");
    }
  };

  const handleError = () => {
    if (!closed) {
      handlers.onError?.("Run event stream disconnected");
    }
  };

  eventSource.addEventListener("run.update", handleUpdate as EventListener);
  eventSource.onerror = handleError;

  return () => {
    closed = true;
    eventSource.removeEventListener("run.update", handleUpdate as EventListener);
    eventSource.close();
  };
}
